### ISSUE TITLE
Now playing not working on grapheneos 

### ISSUE THREAD

**Issue by Alundert** [KieronQuinn/AmbientMusicMod#235]
When installed with shizuku and the ambient music mod working it puts out this error when trying to start:

```
type: crash
osVersion: google/raven/raven:14/AP1A.240305.019.A1/2024030900:user/release-keys
package: com.kieronquinn.app.pixelambientmusic:133
process: com.kieronquinn.app.pixelambientmusic
processUptime: 576 + 91 ms
installer: com.android.packageinstaller

java.lang.UnsatisfiedLinkError: JNI_ERR returned from JNI_OnLoad in "/data/app/~~a69dAWNU0Y_DJ8REdaYJJA==/com.kieronquinn.app.pixelambientmusic-EIhoHTcc3Xo4nMGeIuxrMQ==/lib/arm64/libaliuhook.so"
	at java.lang.Runtime.loadLibrary0(Runtime.java:1082)
	at java.lang.Runtime.loadLibrary0(Runtime.java:1003)
	at java.lang.System.loadLibrary(System.java:1765)
	at de.robv.android.xposed.XposedBridge.<clinit>(XposedBridge.java:29)
	at com.kieronquinn.app.pixelambientmusic.xposed.Xposed.deoptimizeMethod(Xposed.kt:101)
	at com.kieronquinn.app.pixelambientmusic.xposed.XposedHooks.hookMethod(XposedHooks.kt:150)
	at com.kieronquinn.app.pixelambientmusic.xposed.XposedHooks.setupHooks(XposedHooks.kt:145)
	at com.kieronquinn.app.pixelambientmusic.xposed.XposedHooks.init(XposedHooks.kt:109)
	at com.kieronquinn.app.pixelambientmusic.xposed.XposedHooks$Companion.setupHooks(XposedHooks.kt:96)
	at com.kieronquinn.app.pixelambientmusic.Injector.attachBaseContext(Injector.kt:32)
	at android.app.Application.attach(Application.java:352)
	at android.app.Instrumentation.newApplication(Instrumentation.java:1350)
	at android.app.LoadedApk.makeApplicationInner(LoadedApk.java:1467)
	at android.app.LoadedApk.makeApplicationInner(LoadedApk.java:1399)
	at android.app.ActivityThread.handleBindApplication(ActivityThread.java:7299)
	at android.app.ActivityThread.-$$Nest$mhandleBindApplication(Unknown Source:0)
	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2344)
	at android.os.Handler.dispatchMessage(Handler.java:107)
	at android.os.Looper.loopOnce(Looper.java:232)
	at android.os.Looper.loop(Looper.java:317)
	at android.app.ActivityThread.main(ActivityThread.java:8532)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:552)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:894)
```

Is there a way I can fix that?

**Comment by KieronQuinn** [comment]
GrapheneOS is likely blocking the hooks from working due to its enhanced security. This is not something I can fix.

**Comment by KieronQuinn** [comment]
> @KieronQuinn No, we're not. Your app is broken on Android 14 QPR2.

The app has been working on QPR2 on the Pixel 8 Pro and 6 Pro for months. It works fine. It continues to work on QPR3, and while it was initially broken on 15 (due to the stripped symbols breaking LSPlant), this was resolved by switching to another hooking engine for 15 and does not impact 14 (unless you are stripping the symbols also). The error is not part of Ambient Music Mod, it's from Aliucord's hook, which in turn simply loads LSPlant. If you would like to query this further, I suggest you take it up with them as it is not my app to blame.

**Comment by KieronQuinn** [comment]
I'm not going to wade intro drama, I have nothing against GraphemeOS but it was indeed a change to your method signatures which mean they no longer match AOSP that caused this. While not necessarily for security enhancements, it's definitely a GrapheneOS change and not an app bug. I don't appreciate being told flat that I'm apparently lying either. I have therefore hidden your comment and would ask that you correct the record in the linked issue as to not mislead people there either. 

There are two hooking frameworks available that can be used for this app, they're both created by members of the LSposed team, who apparently don't feel like supporting GrapheneOS. Yes, that's a shit situation for us all but there's nothing I can do. I'm not going to create my own just for this app, and doing it without hooking would be extremely difficult or potentially impossible, sorry.

**Comment by KieronQuinn** [comment]
Ok, @thestinger it's reopened. However, I honestly do not see how my comment about the changes to Graphene are not correct. Your own comment says you changed a method to enhance security - what is wrong? 

Is it the "blocking" bit?

**Comment by KieronQuinn** [comment]
Further ping as I'm not sure unlocking actually sends a notification, will delete once not required

**Comment by KieronQuinn** [comment]
Understood. I know it's a hell site but I'll send you a DM about this on Twitter shortly. This message will be deleted after that and no further comment will be made.

**Comment by KieronQuinn** [comment]
This I believe should be inadvertently be fixed in 1.3.4 of Now Playing since it updates hook to include this fix: https://github.com/Aliucord/hook/pull/9, which supports stripped libart. This was the _actual_ issue, and they also are now maintaining a LSPlant fork which should keep it going in the future :)
